service/history/shard/controller_impl.go

```diff
@@ -353,9 +353,17 @@ func (c *ControllerImpl) acquireShards() {
 			c.contextTaggedLogger.Error("Unable to create history shard context", tag.Error(err), tag.OperationFailed, tag.ShardID(shardID))
 			return
 		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), shardIOTimeout)
+		defer cancel()
+		if err := shard.AssertOwnership(ctx); IsShardOwnershipLostError(err) {
```

Comment:
so the problem here is that `AssertOwnership` takes the shard rwlock, so this puts us back in the situation that we've tried to fix, where a shard that's stuck in another persistence op will block an entire acquire worker (so 10 like that would block all workers). the context timeout doesn't help because `rwlock.Lock` is not cancellable.

I don't see a good alternative, though, since it needs to get the rangeid. one option is to just run `AssertOwnership` in a new goroutine. the better solution is probably to rewrite this entire thing without a fixed number of workers, so shards don't block each other.

Reply:
until we have a complete rewrite of this logic, 10 or more goroutine should be fine, even we have a dead lock issue (solve the deadlock issue first?)